openwrt_api.py

```diff
@@ -0,0 +1,400 @@
+  
+  # Check the input arguments (sanity)
+  if interface not in get_network_interface():
+    raise RepyArgumentError("interface "+ interface + " is not available.")
+
+  if os.path.exists("/proc/net/dev"):
+    dev = safe_open('/proc/net/dev', 'r')
+    for line in dev:
+      if interface in line:
+        data = line.split('%s:' % interface)[1].split()
+        rx_packets, tx_packets = (data[1], data[9])
+        return {"recv": rx_packets, "trans": tx_packets}
+  else:
+    raise FileNotFoundError("Could not find /proc/net/dev!")
+
+def get_network_interface():
+  """
+  <Purpose>
+    Returns a list of available network interfaces.
+
+  <Arguments>
+    None
+
+  <Exceptions>
+    FileNotFoundError is raised if the file does not exist
+
+  <Side Effects>
+    None
+
+  <Returns>
+    The list of strings(interface name).
+  """
+  if os.path.exists("/proc/net/dev"):
+    dev = safe_open("/proc/net/dev", "r").readlines()
+    header_line = dev[1]
+    header_names = header_line[header_line.index("|")+1:].replace("|", " ").split()
+
+    result = []
+    for line in dev[2:]:
+      interface = line[:line.index(":")].strip()
+      result.append(interface)
+    return result
+  else:
+    raise FileNotFoundError("Could not find /proc/net/dev!")
+
+def traceroute(dest_ip,port,max_hops,waittime,ttl):
+  """
+  <Purpose>
+    Print the route packets take to network host 
+
+  <Arguments>
+    dest_ip: The ip address to traceroute
+    port: The port to traceroute
+    max_hops: Maximum number of hops
+    waittime: Set the time to wait for a response to a probe
+    ttl: Set what TTL to start.
+
+  <Exceptions>
+    RepyArgumentError when the arguments aren't valid types or value
+    or values
+
+    ResourceForbiddenError when the local port isn't allowed
+
+
+  <Resource Consumption>
+    This operation consumes 28 bytes + number of bytes of the message that
+    were transmitted. This requires that the localport is allowed.
+
+  <Side Effects>
+    None
+
+  <Returns>
+    The result of traceroute
+  """
+  if type(dest_ip) is not str:
+    raise RepyArgumentError("Provided dest_ip must be a string!")
+
+  if type(port) is not int:
+    raise RepyArgumentError("Provided port must be an integer!")
+
+  if type(max_hops) is not int:
+    raise RepyArgumentError("Provided max_hops must be an integer!")
+
+  if type(waittime) is not int:
+    raise RepyArgumentError("Provided waittime must be an integer!")
+
+  if type(ttl) is not int:
+    raise RepyArgumentError("Provided ttl must be an integer!")
+
+  # Check the input arguments (sanity)
+  if not emulcomm._is_valid_ip_address(dest_ip):
+    raise RepyArgumentError("Provided dest_ip is not valid! IP: '" + dest_ip + "'")
+
+  if not emulcomm._is_valid_network_port(port):
+    raise RepyArgumentError("Provided port is not valid! Port: "+str(port))
+
+  if not emulcomm._is_allowed_localport("UDP", port):
+    raise ResourceForbiddenError("Provided port is not allowed! Port: "+str(port))
```

Comment:
`port` is the *destination* port on the remote end, right?. Repy/`emulcomm` restricts *source* ports only. Source and destination ports need not be the same (although in your implementation they seem to be.)